src/cpp/src/whisper/streamer.hpp

```diff
@@ -20,5 +24,122 @@ class ChunkTextCallbackStreamer : private TextCallbackStreamer, public ChunkStre
 };
 
+class WhisperStreamer {
+public:
+    WhisperStreamer(ChunkStreamerVariant& streamer, Tokenizer& tokenizer) {
+        if (auto streamer_obj = std::get_if<std::monostate>(&streamer)) {
+            m_streamer_ptr = nullptr;
+        } else if (auto streamer_obj = std::get_if<std::shared_ptr<ChunkStreamerBase>>(&streamer)) {
+            m_streamer_ptr = *streamer_obj;
+        } else if (auto callback = std::get_if<std::function<bool(std::string)>>(&streamer)) {
+            m_streamer_ptr = std::make_shared<ChunkTextCallbackStreamer>(tokenizer, *callback);
+        }
+    };
+
+    void start() {
+        if (!m_streamer_ptr) {
+            return;
+        }
+
+        m_worker_thread = std::make_shared<std::thread>(&WhisperStreamer::_worker, this);
+    }
+
+    void put_chunk(const std::vector<int64_t>& tokens) {
+        if (!m_streamer_ptr) {
+            return;
+        }
+
+        std::lock_guard<std::mutex> lock(m_mutex);
+        m_queue.push(tokens);
+        m_cv.notify_one();
+    }
+
+    void put(const int64_t token) {
+        if (!m_streamer_ptr) {
+            return;
+        }
+
+        std::lock_guard<std::mutex> lock(m_mutex);
+        m_queue.push(token);
+        m_cv.notify_one();
+    }
+
+    void end() {
+        if (!m_streamer_ptr) {
+            return;
+        }
+
+        {
+            std::lock_guard<std::mutex> lock(m_mutex);
+            m_stopped = true;
+        }
+
+        m_cv.notify_one();
+
+        if (m_worker_thread && m_worker_thread->joinable()) {
+            m_worker_thread->join();
+        }
+
+        std::lock_guard<std::mutex> lock(m_mutex);
+        m_streamer_ptr->end();
+    }
+
+    bool is_dropped() {
+        std::lock_guard<std::mutex> lock(m_mutex);
+        return m_dropped;
+    }
+
+private:
+    std::shared_ptr<ChunkStreamerBase> m_streamer_ptr = nullptr;
+    std::shared_ptr<std::thread> m_worker_thread = nullptr;
+    std::mutex m_mutex;
+    std::condition_variable m_cv;
+    std::queue<std::variant<int64_t, std::vector<int64_t>>> m_queue;
+
+    bool m_stopped = false;
+    bool m_dropped = false;
+
+    void _worker() {
+        while (true) {
+            std::variant<int64_t, std::vector<int64_t>> token_variant;
+            {
+                std::unique_lock<std::mutex> lock(m_mutex);
+
+                // wait for the next token in queue or if streamer was stopped
+                m_cv.wait(lock, [this] {
+                    return m_stopped || !m_queue.empty();
+                });
+
+                // continue streaming until queue is empty
+                if (m_stopped && m_queue.empty()) {
+                    break;
+                }
+
+                token_variant = m_queue.front();
+                m_queue.pop();
+            }
+
+            // wait for streamer_ptr result
+            bool is_dropped = false;
+
+            if (auto token = std::get_if<int64_t>(&token_variant)) {
+                is_dropped = m_streamer_ptr->put(*token);
```

Comment:
what if this callback is never return? e.g. it sleeps for a while (or at least it latest longer than generation of output tokens) 

Looks like pipeline will not wait for streaming completion and will go out of WhiserPipeline::generate and even ~WhiserPipeline can be called on app exit, but streaming it will waiting.

Reply:
If callback will sleep (or never return) pipeline will not exit and will wait for it here: https://github.com/openvinotoolkit/openvino.genai/pull/1642/files#diff-9eb3f21f8f27be6218b861f0159dc50546537f1bef39eb1a4ecf9b626ec8918cR357